### ISSUE TITLE
determine stanc3 distribution 

### ISSUE THREAD

**Issue by mitzimorris** [stan-dev/cmdstan#797]
#### Summary:
The stanc3 compiler is distributed as a pre-built binary - determine the best way to include this in the release and change the release procedures accordingly.

#### Description:
https://discourse.mc-stan.org/t/releasing-stanc3-in-cmdstan-options/12679

**Comment by mitzimorris** [comment]
@serban-nicusor-toptal - what is going to happen with the release?

**Comment by wds15** [comment]
I think it would be preferable if downloading cmdstan does the job. So no more online connection should be needed after having downloaded cmdstan.

**Comment by serban-nicusor-toptal** [comment]
Hey, @mitzimorris I think it's best to go with what @wds15 said, the simplest solution for the user and includes fewer steps.

I have the following questions for the release:

1. How and where should the above be implemented? If it's done in the makefile of cmdstan it won't change the release process?
2. What are the next release versions? Last time it was a bit tricky with 2.21.0 and 3.0
3. Last time I've received a very [well made](https://github.com/stan-dev/math/releases/tag/v3.0.0) release readme, will someone have time to build one for the next version? I think it's way better than the simple static templates I use because it's from a developer to a developer perspective.
4. When should I do the release?
5. Any information or change that I should be aware of before doing the release or docs.

I wrote a small Jenkin job locally that I've been playing with to build docs then create a PR with them in the docs repo. I will add it to the live Jenkins and ask you to review if you have the time, also I don't know how to change the versions in [here](https://mc-stan.org/users/documentation/) after generating docs. I think you've already told me once but I forgot.

Thanks!

**Comment by mitzimorris** [comment]
@serban-nicusor-toptal - are the stanc3 tagged with the same version as the rest of the release?   if so, the solution is:

- include the release version of the stanc3 pre-built binaries as part of the cmdstan tarballs.
- change the makefile, which currently download the nightly release of stanc3 - definitely not what we want to do - and instead, have it copy (or symlink?) the os-specific binary with bin/stanc.

I can fix the makefile and will submit a PR if this is the way we want to go.

**Comment by serban-nicusor-toptal** [comment]
Looks like the today release of stanc3 is tagged with the proper [version](https://github.com/stan-dev/stanc3/releases)
I'd go for copying it since windows doesn't like symlinks.

I'm in for this way of doing it, what do you think @seantalts @rok-cesnovar @jgabry ?

**Comment by mitzimorris** [comment]
> Hey, @mitzimorris I think it's best to go with what @wds15 said, the simplest solution for the user and includes fewer steps.
> 
> I have the following questions for the release:
> 
> 1. How and where should the above be implemented? If it's done in the makefile of cmdstan it won't change the release process?

see above comment - if we get the stanc3 versioning and the cmdstan makefile fixed, this won't change the release process.

> 2. What are the next release versions? Last time it was a bit tricky with 2.21.0 and 3.0, but now I know docs should be pinned to Stan version.

the Stan version is 2.22.0 - it's a version 2 release because we're preserving backwards compatibility with all existing features in the language (AFAIK?)

> 3. Last time I've received a very [well made](https://github.com/stan-dev/math/releases/tag/v3.0.0) release readme, will someone have time to build one for the next version? I think it's way better than the simple static templates I use because it's from a developer to a developer perspective.

agreed, that readme is very nice.

> 4. When should I do the release?

the CmdStan release can be done anytime after the Stan and stanc3 release.

> 5. Any information or change that I should be aware of before doing the release or docs.

@rok-cesnovar - any features added?  bugfixes that I know of are json parser now accepts multi-dim arrays of size 0,  also, there was a makefile bugfix - will track down PRS.

> I wrote a small Jenkin job locally that I've been playing with to build docs then create a PR with them in the docs repo. I will add it to the live Jenkins and ask you to review if you have the time, also I don't know how to change the versions in [here](https://mc-stan.org/users/documentation/) after generating docs. I think you've already told me once but I forgot.
> 

`make manual` - this builds the CmdStan User's Guide pdf in directory `doc`

> Thanks!

You're most welcome - thank you!

**Comment by rok-cesnovar** [comment]
This is it on Cmdstan side: 

- Cmdstan now prints info on threading and OpenCL when either are used
- Cmdstan now uses pre-built binary of the Stan compiler written in OCaml
- add a new make rule for cleaning up compiled Stan programs

Bugfix: 
- accepts arrays where any dimension is size 0
- cleanup use of Eigen vectors of strings
- makefiles fixes for cases where users delete the model .hpp file

Developer oriented:
- fixed script for running tests
- fixed dependencies and makefiles which now recompile on changes in services
- makefile fixes for cases when changes were done in the stan submodule
- cleanup of hardcoded version numbers

I will try to write similar release notes as for 2.21 for Stan Math with Steve's script.

**Comment by rok-cesnovar** [comment]
Here is my quick draft for the Math release notes: https://gist.github.com/rok-cesnovar/923fcc78139f4331fcb73c28afc7a332#file-math-3-1-release-notes-draft

Hopefully I included everything. Would be great to have another set of eyes on it. Tagging @mcol, @SteveBronder, @t4c1 but anyone is welcome.

**Comment by rok-cesnovar** [comment]
Stan Math release notes are now a bit more polished https://gist.github.com/rok-cesnovar/923fcc78139f4331fcb73c28afc7a332#gistcomment-3145261

cc: @serban-nicusor-toptal 

For Stan there weren't many changes. I think this is it:

Features: 
- validate_dims in stan::io::var_context vas made a virtual function which allows its implementation for JSON data which is needed to accepts arrays where any dimension is size 0
- replaced Eigen vectors of strings with std::vector<std::string>
- replaced stan::math::domain_error with throw_domain_error
- added const noexcept accessors to the stepsize adaptation class which allows developers to query the adaptation parameters in contexts where it's natural to have a const sampler (logging, serialization, ...)

**Comment by wds15** [comment]
The biggest change is - to me - switching to the new stanc3 parser. That needs to be clearly visible.

**Comment by rok-cesnovar** [comment]
Yes, I agree. That is noted in the Cmdstan notes draft [here](https://github.com/stan-dev/cmdstan/issues/797#issuecomment-575907260). We can move it to the top and bold it. 

Though users actually wont even notice it, except for the features/bug fixes noted in the stanc3 release notes. For now its supposed to replicate the existing compiler.

**Comment by wds15** [comment]
Users will notice. The error messages are different and for complicated Stan programs its going to be interesting. I know that from my own ones (don't get me wrong - no open issue known to me).

**Comment by mitzimorris** [comment]
> We can move it to the top and bold it.

absolutely.



> Users will notice.

double absolutely.

**Comment by mitzimorris** [comment]
this issue isn't quite the right place to raise the question, are we ready to release the stanc3 compiler?

**Comment by wds15** [comment]
I think we have to bite the bullet at some point - so why not now?

I agree with you that users should know where to turn to if they find trouble. However, it will be hard for users to discern if something is a `stanc3` or a cmdstan issue. So we could just let everyone file stuff here for now and move it over? I think moving issue between repos is easy by now; I could be wrong.

**Comment by serban-nicusor-toptal** [comment]
Hey @mitzimorris can you please review the docs generation job when you have a min? Thank you.
If you have any suggestions of what we can improve or add to remove manual work related to it that would be nice.

Jenkinsfile can be found [here](https://github.com/stan-dev/ci-scripts/blob/refactor-jenkins-docs/docs/Jenkinsfile), after feedback and changes I will push it in docs master branch.
Visual representation in Blue Ocean [here](https://jenkins.mc-stan.org/blue/organizations/jenkins/BuildDocs/detail/BuildDocs/14/pipeline)
Just creates a [branch](https://github.com/stan-dev/docs/tree/docs-2-22) and then a [PR](https://github.com/stan-dev/docs/pull/136) for it.
It also clones cmdstan and builds the manual, you can find it attached in the job artifacts [here](https://jenkins.mc-stan.org/job/BuildDocs/14/).

**Comment by rok-cesnovar** [comment]
> packaging

I think the majority opinion was for packaging three OS-specific packages with binaries included. That way `make build` wont download latest nightly stanc3 build which it does currently.

The only question that remains if we want to change what happens if someone deletes the stanc executable. Does `make build` download the nightly build or the release build. If we go with the latter we should then also add a ˙make build-nightly` or something along those lines.

> user-facing documentation

Changes are listed here https://github.com/stan-dev/stanc3/wiki/changes-from-stanc2

> * how to help test and debug?

Not sure on this one but I guess open an issue with the minimal example as usual. Or was something else meant here.

**Comment by bob-carpenter** [comment]
Can we have an option to pull down the parser version that matches a release?

**Comment by rok-cesnovar** [comment]
Yes, thats what i mean by “download the release compiler/parser build”. Make build currently pulls https://github.com/stan-dev/stanc3/releases/tag/nightly but we could also use https://github.com/stan-dev/stanc3/releases/tag/v2.22.0

**Comment by wds15** [comment]
<spoil>
since everyone managing this release is on this PR (sorry to spoil)... could someone kick of performance tests for this:
https://github.com/stan-dev/math/pull/1632
It could be worthwhile to wait to get this in. It will speedup *all* models by saving `chain` calls for all declared parameters. That could be noticable in models depending on the remaining complexity, etc.

</spoil>

**Comment by mitzimorris** [comment]
> Hey @mitzimorris can you please review the docs generation job when you have a min? Thank you.
> If you have any suggestions of what we can improve or add to remove manual work related to it that would be nice.

hi @serban-nicusor-toptal - looks good to me - many thanks!

**Comment by serban-nicusor-toptal** [comment]
Is there anything else on the checklist for the release ? Can I proceed with the release of `2.22` for Stan, CmdStan and stanc3. Also math becomes `3.1` right ?

**Comment by rok-cesnovar** [comment]
Math will be 3.1 yes. I think Math and Stan are ready, for cmdstan we should wait for https://github.com/stan-dev/cmdstan/pull/799 and for stanc3 we should wait for https://github.com/stan-dev/stanc3/issues/435 (I think this one was already fixed) and https://github.com/stan-dev/stanc3/issues/436

**Comment by mitzimorris** [comment]
CmdStan release notes always just say "see Stan and Math release notes" - we need to document the new parser, per discussion https://github.com/stan-dev/stanc3/issues/436#issuecomment-576613834 

@rok-cesnovar - happy to help with release notes - where are they?

**Comment by rok-cesnovar** [comment]
We drafted a math release notes here https://gist.github.com/rok-cesnovar/923fcc78139f4331fcb73c28afc7a332#file-math-3-1-release-notes-draft

Not much happened for Stan so this is it here https://github.com/stan-dev/cmdstan/issues/797#issuecomment-576204588

**Comment by rok-cesnovar** [comment]
Ok, so docs on how to diagnose potential stanc3 issues and we are ready as far as I understood.

I do have one question that was pointed out in a PR.

@mitzimorris wrote: 

> the intention is that the release tarball includes all OS-specific releases - because the tarball is platform agnostic - this means that the scripts that create the tarball need to include the tagged os-specific binaries for stanc3 - @serban-nicusor-toptal - is this doable?

We could also do platform-specific tarballs or could do both: a platform agnostic tarball and more lightweight platform specific tarballs (potentially a support nightmare due to confusion). The downside of platform agnostic is the added 30+ MB (not really a big big deal) of binaries while the downside of going platform specific is the potential confusion and additional work for cmdstanX interfaces. I am fine either way, just pointing this out.

**Comment by mitzimorris** [comment]
OK, drafting docs on stanc3 - will have something ready this afternoon.

I'd prefer a platform agnostic tarball - note that the CmdStanX wrappers are downloading this one, so adding platform-dependent versions will propogate complexity to the wrappers as well.

**Comment by rok-cesnovar** [comment]
I am actually close to finishing a PR to cleanup the manual/readme so you could wait on that to see if there is anything else to add.

**Comment by rok-cesnovar** [comment]
Great, I think we are all set now for the release.

**Comment by rok-cesnovar** [comment]
Closing this issue as I think this was settled and we can continue any release related discussion on the release notes issue.
